python/pyarrow/tests/test_udf.py

```diff
@@ -0,0 +1,458 @@
+                                {},
+                                pa.int64())
+
+    res = pc.call_function("random_func", [])
+    assert res.as_py() >= 0 and res.as_py() <= 10
+
+
+def test_output_datatype():
+    def add_one(ctx, array):
+        ar = pc.call_function("add", [array, 1])
+        ar = ar.cast(pa.int32())
+        return ar
+
+    func_name = "py_add_to_scalar"
+    in_types = {"array": pc.InputType.array(pa.int64())}
+    out_type = pa.int64()
+    doc = {
+        "summary": "add function scalar",
+        "description": "add function"
+    }
+    pc.register_scalar_function(add_one, func_name, doc,
+                                in_types, out_type)
+
+    func = pc.get_function(func_name)
+
+    assert func.name == func_name
+
+    expected_expr = "Expected output type, int64," \
+        + " but function returned type int32"
+
+    with pytest.raises(pa.lib.ArrowTypeError, match=expected_expr):
+        pc.call_function(func_name, [pa.array([20, 30])])
+
+
+def test_output_value():
+    def add_one(ctx, array):
+        ar = pc.call_function("add", [array, 1])
+        ar = ar.cast(pa.int32())
+        return ar
+
+    func_name = "test_output_value"
+    in_types = {"array": pc.InputType.array(pa.int64())}
+    out_type = {}
+    doc = {
+        "summary": "test output value",
+        "description": "test output"
+    }
+
+    expected_expr = "DataType expected, got <class 'dict'>"
+
+    with pytest.raises(TypeError, match=expected_expr):
+        pc.register_scalar_function(add_one, func_name, doc,
+                                    in_types, out_type)
+
+
+def test_output_type():
+    def add_one(ctx, array):
+        return 42
+
+    func_name = "add_to_scalar_as_py"
+    in_types = {"array": pc.InputType.array(pa.int64())}
+    out_type = pa.int64()
+    doc = {
+        "summary": "add function scalar",
+        "description": "add function"
+    }
+    pc.register_scalar_function(add_one, func_name, doc,
+                                in_types, out_type)
+
+    func = pc.get_function(func_name)
+
+    assert func.name == func_name
+
+    expected_expr = "Unexpected output type: int"
+
+    with pytest.raises(pa.lib.ArrowTypeError, match=expected_expr):
+        pc.call_function(func_name, [pa.array([20, 30])])
+
+
+def test_input_type():
+    def add_one(ctx, array):
+        return 42
+
+    func_name = "test_input_type"
+    in_types = {"array": None}
+    out_type = pa.int64()
+    doc = {
+        "summary": "test invalid input type",
+        "description": "invalid input function"
+    }
+    expected_expr = "in_types must be of type InputType"
+
+    with pytest.raises(TypeError, match=expected_expr):
+        pc.register_scalar_function(add_one, func_name, doc,
+                                    in_types, out_type)
+
+
+def test_udf_context():
+
+    def random(context, one, two):
+        return pc.add(one, two, memory_pool=context.memory_pool)
```

Comment:
This probably won't exercise the memory pool since this test is adding two numeric scalars.

Reply:
Thanks a lot @pitrou much appreciated. But let me work and learn a few things from you all, at least for this PR :) . 
I am thinking of a better example for this. May be leverage the batch_length and write one. 
I understand this test is not helping much for a user who is familiar with the API and wants to learn a thing or two.

Reply:
Do you have more suggestions in mind to expand the test cases? I can work on them.

Reply:
IIRC you cannot actually set the batch_length manually. It only happens in the context of the query engine.

Reply:
You can check that the batch_length matches the actual inputs.
As for the memory pool, you can instantiate a proxy memory pool and check that the stats are updated in the right pool when allocating.

Reply:
Yeah we can do that. Let's settle for that, it's neat.

Reply:
The maximum batch length is fixed and not configurable but if your dataset's batches are small enough then your batch length should vary with input.

Reply:
Yes, we can leverage it here, right. Say we just pass 3 objects, then the `batch_length` should be 3 right?

Reply:
Yes, I think it's probably a safe bet that the max batch length will always be at least 1000.

Reply:
I see. Noted. Let me work on this.

Reply:
@pitrou I have updated the test cases with fixtures. Could you check if it has the intended use? 
Also I have updated the context usage example. 

cc @westonpace @lidavidm

Reply:
That's not what this is about, right? This is about having a "length" when all arguments are scalar.